Use numeric sorting if possible in project panel

mrnugget opened the pull request:
Previously, if you had the following files/folders in your project

    1-abc
    10
    11-def
    ...
    2
    21-abc

that's how we'd display them.

With this change, we now try to parse them as numbers, if possible, and use that to sort. If we can't parse a component as a number, we fall back to normal string comparison.

End result is this:

    1-abc
    2
    10
    11-def
    ...
    21-abc


Release Notes:

- Fixed filenames with numeric components (`1.txt`, `1/one.txt`, ...) not being sorted as numbers, but as string.

Before:

![screenshot-2024-02-27-18 29 43@2x](https://github.com/zed-industries/zed/assets/1185253/2d223126-329f-4ae7-9a12-d33e2c3fe52f)


After:
![after](https://github.com/zed-industries/zed/assets/1185253/f4f98fa0-e66f-40aa-aa28-189143cbb75f)

mrnugget committed:
Use numeric sorting if possible in project panel

Previously, if you had the following files/folders in your project

  1
  10
  11
  ...
  2
  21

that's how we'd display them.

With this change, we now try to parse them as numbers, if possible, and
use that to sort. If we can't parse a component as a number, we fall
back to normal string comparison.

End result is this:

  1
  2
  10
  11
  ...
  21

maxdeviant commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -1178,11 +1178,25 @@ impl ProjectPanel {
+                                let b_as_number = Path::new(component_b.as_os_str())
+                                    .file_stem()
+                                    .and_then(|s| s.to_str())
+                                    .map(|stem| stem.parse::<i32>());
```
Is it true that right now this only works for file stems consisting solely of numbers?

I wonder if we should try to detect if the file stem starts with a sequence of digits and then sort based on the numeric representation of that first, and then sort by the remaining string representation.

I’m thinking of files like migration names that could start with a numeric timestamp and then have a `-migration-name` suffix.

mrnugget replied:
You're right, that's a good point. Let me take a stab at that.

mrnugget replied:
Implemented together in b4386cab2

mrnugget committed:
Use possible numeric prefix for sorting filenames

This changes the behavior to sort files based on numeric *prefixes*, not
just their numeric file stem.

Previously we would sort files like these incorrectly:

    1-abc.txt
    11-abc.txt
    2-abc.txt
    20-abc.txt

But with this change here, we now sort them correctly:

    1-abc.txt
    2-abc.txt
    11-abc.txt
    20-abc.txt

Co-authored-by: Marshall <marshall@zed.dev>

mrnugget commented:
No. In the scope of this ticket, we only changed the pre-existing sort and didn't extend it any further than that.
